docs/content/preview/architecture/transactions/concurrency-control.md

```diff
@@ -0,0 +1,1226 @@
+<br>
+
+```output
+ k | v
+---+---
+ 1 | 1
+(1 row)
+```
+
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>
+   </td>
+   <td>
+
+```sql
+select * from test;
+```
+
+```output
+ERROR:  Operation expired: Heartbeat:
+Transaction 13fb5a0a-012d-4821-ae1d-5f7780636dd4 expired
+or aborted by a conflict: 40001
+```
+
+```sql
+rollback;
+```
+
+   </td>
+  </tr>
+  <tr>
+   <td>
+
+```sql
+commit;
+```
+
+   </td>
+   <td>
+   </td>
+  </tr>
+</tbody>
+</table>
+
+### Die example
+<table class="no-alter-colors">
+<thead>
+    <tr>
+    <th>
+    Client 1
+    </th>
+    <th>
+    Client 2
+    </th>
+    </tr>
+  </thead>
+  <tbody>
+  <tr>
+   <td>
+   </td>
+   <td>
+
+```sql
+SET yb_transaction_priority_lower_bound = 0.6;
+```
+<br>
+
+   </td>
+  </tr>
+  <tr>
+   <td>
+
+```sql
+SET yb_transaction_priority_upper_bound = 0.4;
+```
+<br>
+
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>
+   </td>
+   <td>
+
+```sql
+begin transaction isolation level repeatable read;
+```
+<br>
+
+   </td>
+  </tr>
+  <tr>
+   <td>
+   </td>
+   <td>
+
+```sql
+select * from test where k=1 for update;
+```
+<br>
+
+```output
+ k | v
+---+---
+ 1 | 1
+(1 row)
+```
+   </td>
+  </tr>
+  <tr>
+   <td>
+
+```sql
+begin transaction isolation level repeatable read;
+```
+<br>
+
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>
+
+```sql
+select * from test where k=1 for update;
+```
+<br>
+
+```output
+ERROR:  All transparent retries exhausted. could not serialize
+access due to concurrent update
+```
+
+```sql
+rollback;
+```
+
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>
+   </td>
+   <td>
+
+```sql
+commit;
+```
+
+   </td>
+  </tr>
+</tbody>
+</table>
+
+### Best-effort internal retries for first statement in a transaction
+
+Note that we see the error message `All transparent retries exhausted` in the preceding example because if the transaction T1, when executing the first statement, finds another concurrent conflicting transaction with equal or higher priority, then T1 will perform a few retries with exponential backoff before giving up in anticipation that the other transaction will be done in some time. The number of retries are configurable by the `ysql_max_write_restart_attempts` YB-TServer gflag and the exponential backoff parameters are the same as the ones described in [Performance tuning](../read-committed/#performance-tuning).
+
+Each retry will use a newer snapshot of the database in anticipation that the conflicts might not occur (if an earlier conflicting transaction T2 has committed with a commit time before the read time of the new snapshot, the conflicts with T2 would essentially be voided since T1 and T2 would no longer be "concurrent").
+
+Note that the retries will not be performed in case the amount of data to be sent from YSQL to the client proxy exceeds the TServer gflag `ysql_output_buffer_size`.
+
+## Wait-on-Conflict [[Beta]](/preview/faq/general/#what-is-the-definition-of-the-beta-feature-tag) {#wait-on-conflict}
```

Comment:
The `{#wait-on-conflict}` is a manually-added anchor. Don't remove it, and links to this heading will stay working even after we remove the beta tag.